src/tools/simeng/gdb/GDBStub.cc

```diff
@@ -0,0 +1,331 @@
+#include "simeng/GDBStub.hh"
+
+namespace simeng {
+
+GDBStub::GDBStub(simeng::Core& core, simeng::MemoryInterface& dataMemory)
+    : core_(core), dataMemory_(dataMemory) {}
+
+int GDBStub::run() {
+  // verbose_ = 1; // uncomment this if you'd like the GDBStub to run verbosely
+
+  int connection = openSocket(2424);  // change this if the port is blocked
+
+  std::cout << "Connection to GDB client established, debugging in progress"
+            << std::endl
+            << std::endl;
+
+  char buffer[10000];
+
+  while (true) {
+    read(connection, buffer, 10000);
+
+    // '+' is an acknowledgement of successful receipt of message
+    // TODO: handle '-', no acknowledgement
+    std::regex ack_regex("^\\+.*");
+
+    // First parentheses = package, second = checksum
+    std::regex packet_regex("^\\$([^#]*)#([0-9a-f]{2})");
+    std::smatch packet_match;
+    std::string bufferString = buffer;
+
+    if (regex_match(bufferString, ack_regex)) {
+      if (verbose_)
+        std::cout << CYAN << "<- Received message acknowledgement" << std::endl
+                  << RESET;
+    } else if (regex_match(bufferString, packet_match, packet_regex)) {
+      std::string packet = packet_match[1].str();
+      if (verbose_) {
+        std::cout << CYAN;
+        if (packet_match.size() == 3)
+          std::cout << "<- Received packet: " << packet << std::endl;
+        else
+          std::cout << RED << "<- Unknown message"
+                    << std::endl;  // TODO: take action
+        std::cout << RESET;
+      }
+      std::string checksum = packet_match[2].str();
+      std::string expected = computeChecksum(packet);
+      if (verbose_ && checksum != expected) {
+        std::cout << RED << "   Checksum failed\n"
+                  << "   Expected checksum: " << expected
+                  << "   , but received: " << checksum << std::endl
+                  << RESET;
+        // TODO: take action
```

Comment:
What should be the expected action for a failed checksum? i.e. abort and exit entirely?

Reply:
this was easy enough to find out - respond with a "-" packet - fixed in a commit